### PR TITLE
[backends] Finish CompiledFunction refactoring

### PR DESCRIPTION
**Pull request by bertmaher**
I think this is the light at the end of the tunnel for #1234

### CODE REVIEW

File: include/glow/ExecutionEngine/ExecutionEngine.h [hunk 1]
```diff
@@ -45,6 +46,8 @@ class ExecutionEngine final {
   std::unique_ptr<Module> M_;
   /// The network execution backend.
   std::unique_ptr<Backend> backend_;
+  /// A glow function compiled for this ExecutionEngine's backend.
+  std::unique_ptr<CompiledFunction> function_;
```
Comment (rdzhabarov) [inline include/glow/ExecutionEngine/ExecutionEngine.h]: we should probably reset this function on setBackend() call.

File: tests/unittests/BackendTest.cpp [hunk 2]
```diff
@@ -145,8 +145,8 @@ TEST_P(BackendTest, debugPrint) {
   IRBuilder(IR.get()).createDebugPrintInst("print", *IR->getWeights().begin());
 
   std::unique_ptr<Backend> backend(createBackend(GetParam()));
-  backend->init(std::move(IR));
-  backend->doForwardPass();
+  auto function = backend->compile(std::move(IR));
```
Comment (jfix71) [inline tests/unittests/BackendTest.cpp]: nit: I prefer explicitly using * for auto pointers

File: include/glow/Backends/Backend.h [hunk 3]
```diff
@@ -44,29 +39,26 @@ class Backend {
...
 
-  /// Prepare the interpreter for execution of new code.
-  virtual void init(std::unique_ptr<IRFunction> IR) = 0;
+  /// Generate code for input function \param IR.
+  virtual std::unique_ptr<CompiledFunction>
+  compile(std::unique_ptr<IRFunction> IR) const = 0;
 
   /// Save the bundle for a later standalone execution.
-  virtual void save(std::unique_ptr<IRFunction> IR, llvm::StringRef outputDir) {
+  virtual void save(std::unique_ptr<IRFunction> IR, llvm::StringRef outputDir) const {
```
Comment (jfix71) [inline include/glow/Backends/Backend.h]: nit: make sure you clang-format!

File: tests/unittests/BackendTest.cpp (hunk 2 above)
Comment (bertmaher) [reply tests/unittests/BackendTest.cpp]: It's a `unique_ptr` though, so I don't think `auto *` is correct.

**Pull request merged**